File: 11_动态规划/dp分类/区间dp/dfs/回文/516. 最长回文子序列.py
from functools import lru_cache


class Solution:
    def longestPalindromeSubseq(self, s: str) -> int:
        """最长回文子序列"""

        @lru_cache(None)
        def dfs(left: int, right: int) -> int:
            if left > right:
                return 0
            if left == right:
                return 1
            if s[left] == s[right]:
                return dfs(left + 1, right - 1) + 2
            return max(dfs(left, right - 1), dfs(left + 1, right))

        n = len(s)
        return dfs(0, n - 1)


print(Solution().longestPalindromeSubseq(s="bbbab"))

File: 11_动态规划/dp分类/区间dp/dfs/回文/516. 最长回文子序列.ts
/** 最长回文子序列 (LPS). */
function longestPalindromeSubsequence(arr: ArrayLike<unknown>): number {
  const n = arr.length
  const dp = new Uint16Array(n * n)
  for (let i = n - 1; ~i; i--) {
    dp[i * n + i] = 1
    for (let j = i + 1; j < n; j++) {
      if (arr[i] === arr[j]) {
        dp[i * n + j] = dp[(i + 1) * n + j - 1] + 2
      } else {
        dp[i * n + j] = Math.max(dp[(i + 1) * n + j], dp[i * n + j - 1])
      }
    }
  }
  return dp[n - 1]
}

export { longestPalindromeSubsequence }

if (require.main === module) {
  console.log(longestPalindromeSubsequence('bbbab'))
  // https://leetcode.cn/problems/longest-palindromic-subsequence/
  // eslint-disable-next-line no-inner-declarations
  function longestPalindromeSubseq(s: string): number {
    return longestPalindromeSubsequence(s)
  }
}
